### PR TITLE
Added Gemini Provider

### LINKED ISSUE #19: gemini support
can you just provide an environment variable?

maybe this should be an extension, using pluggy or similar.

### PR DESCRIPTION
**Pull request by Siddhesh-Agarwal**
- Added Gemini Provider
  - Added API key to env template and `Settings` class
  - closes #19 
- Improved type hinting across all files

## Summary by CodeRabbit

- **New Features**
	- Introduced the `Gemini` provider for enhanced API interaction.
	- Added a new `LoggingConfig` class for structured logging configuration.

- **Bug Fixes**
	- Improved error handling in various provider classes to ensure clearer messages and robust functionality.

- **Documentation**
	- Updated `README.md` for clarity and restructured supported APIs.

- **Refactor**
	- Enhanced type annotations across various classes and methods for better clarity and type safety.
	- Changed property decorators to `@cached_property` for improved performance in multiple provider classes.

- **Chores**
	- Updated `pyproject.toml` to lower the minimum required Python version and added a new dependency for `google-generativeai`.

### CODE REVIEW

**Comment by Siddhesh-Agarwal** [comment]
- Changed base Python version to 3.10 to close #21 
- Added dependency in `pypoetry.toml`

File: README.md [hunk 1]
```diff
@@ -49,7 +50,7 @@ First, authenticate your API keys by setting them in the environment variables:
 $ export OPENAI_API_KEY="sk-..."
 ```
 
-This pattern allows you to keep your API keys private and out of your codebase. Other supported environment variables: `ANTHROPIC_API_KEY`, `XAI_API_KEY`, and `GROQ_API_KEY`.
+This pattern lets you keep your API keys private and out of your codebase. Other supported environment variables: `ANTHROPIC_API_KEY`, `XAI_API_KEY`, and `GROQ_API_KEY`.
```
Comment (kennethreitz) [inline README.md:53]: can we keep this as it was? thank you

File: README.md [hunk 2]
```diff
@@ -140,7 +141,7 @@ response = gpt_4o_mini.generate_text(
 Harnessing the power of Python, you can easily create your own plugins to add additional functionality to your conversations:
 
 ```python
-class SimpleMemoryPlugin:
+class SimpleMemoryPlugin(sm.BasePlugin):
```
Comment (kennethreitz) [inline README.md:144]: i also want to keep the simple memory plugin just a base class, to show that it's allowed.

File: simplemind/__init__.py [hunk 3]
```diff
@@ -77,7 +77,12 @@ def create_conversation(
 
 
 def generate_data(
-    prompt, *, llm_model=None, llm_provider=None, response_model=None, **kwargs
+    prompt: str,
+    *,
+    llm_model: str | None = None,
```
Comment (kennethreitz) [inline simplemind/__init__.py:82]: why is this better than Optional?

File: simplemind/models.py [hunk 4]
```diff
@@ -22,34 +23,41 @@ def __repr__(self):
...
 
-class BasePlugin:
+class BasePlugin(SMBaseModel, ABC):
     """The base conversation plugin class."""
 
     # Plugin metadata.
     meta: Dict[str, Any] = {}
 
-    def initialize_hook(self, conversation: "Conversation"):
+    @abstractmethod
```
Comment (kennethreitz) [inline simplemind/models.py:32]: please remove the @abstractmethods, as they require all of them to be implemented in the subclass, and someone may want only one

File: simplemind/models.py [hunk 5]
```diff
@@ -73,7 +72,7 @@ class Conversation(SMBaseModel):
     messages: List[Message] = []
     llm_model: Optional[str] = None
     llm_provider: Optional[str] = None
-    plugins: List[Any] = []
+    plugins: List[BasePlugin] = []
```
[the remark this answers is not in the document]
Comment (kennethreitz) [reply simplemind/models.py:95]: this is accurate

File: simplemind/__init__.py (hunk 3 above)
Comment (Siddhesh-Agarwal) [reply simplemind/__init__.py:82]: While they are the same thing, I guess it is better to explicitly state the thing instead of "implying" it via Optional

Comment (kennethreitz) [reply simplemind/__init__.py:82]: ah interesting. not opposed

**Comment by kennethreitz** [comment]
Looking forward to merging this!

File: simplemind/models.py (hunk 4 above)
Comment (Siddhesh-Agarwal) [reply simplemind/models.py]: fixed this

File: README.md (hunk 2 above)
Comment (Siddhesh-Agarwal) [reply README.md:144]: done 👍🏻

File: README.md (hunk 1 above)
Comment (Siddhesh-Agarwal) [reply README.md]: reverted

**Comment by kennethreitz** [comment]
lgtm! ✨🍰✨

**Comment by kennethreitz** [comment]
made some changes, but shipped and released!

**Pull request merged**